### PR TITLE
[risk=low][no ticket] utils/dates.tsx

### PR DESCRIPTION
**Pull request by jmthibault79**
Move a few functions/consts from `utils/index.tsx` to a new `utils/dates.tsx`

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/main/e2e/README.md#examples)
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: ui/src/app/pages/homepage/recent-resources.tsx [hunk 1]
```diff
@@ -136,7 +137,7 @@ export const RecentResources = fp.flow(withCdrVersions())((props: Props) => {
               {getDisplayName(r)}
             </ResourceNavigation>,
             workspaceName: <WorkspaceNavigation workspace={getWorkspace(r)} resource={r} style={styles.navigation}/>,
-            formattedLastModified: formatWorkspaceResourceDisplayDate(r.lastModifiedEpochMillis),
+            formattedLastModified: displayDateWithoutHours(r.lastModifiedEpochMillis),
```
Comment (jmthibault79) [inline ui/src/app/pages/homepage/recent-resources.tsx:140]: identical except for a comma: `Dec 13 2021` becomes `Dec 13, 2021`

File: ui/src/app/components/resource-card.tsx [hunk 2]
```diff
@@ -150,7 +150,7 @@ class ResourceCard extends React.Component<Props, {}> {
           </FlexColumn>
           <div style={styles.cardFooter}>
             <div style={styles.lastModified} data-test-id='last-modified'>
-              Last Modified: {formatWorkspaceResourceDisplayDate(resource.lastModifiedEpochMillis)}</div>
+              Last Modified: {displayDateWithoutHours(resource.lastModifiedEpochMillis)}</div>
```
Comment (jmthibault79) [inline ui/src/app/components/resource-card.tsx:153]: Identical except for a comma: `Last Modified: Nov 15 2021` becomes `Last Modified: Nov 15, 2021`

File: ui/src/app/utils/resources.tsx [hunk 3]
```diff
@@ -110,13 +110,3 @@ export function convertToResource(
     adminLocked
   };
 }
-
-export function formatWorkspaceResourceDisplayDate(time: number): string {
```
Comment (jmthibault79) [inline ui/src/app/utils/resources.tsx:114]: near-duplicate of `displayDateWithoutHours()`

File: ui/src/app/utils/dates.tsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    });
+}
+
+export function displayDateWithoutHours(time: number): string {
+  const date = new Date(time);
+  // datetime formatting to slice off weekday and exact time
+  return date.toLocaleString('en-us', {month: 'short', day: 'numeric', year: 'numeric'});
+}
+
+export function isDateValid(date: Date): boolean {
```
Comment (jmthibault79) [inline ui/src/app/utils/dates.tsx:23]: this function is new

File: ui/src/app/utils/dates.tsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import {isBlank} from './index';
+
+export const MILLIS_PER_DAY = 24 * 60 * 60 * 1000;
+export const daysFromNow = (timeInMillis: number): number => Math.floor((timeInMillis - Date.now()) / MILLIS_PER_DAY);
+
+// To convert datetime strings into human-readable dates
+export function displayDate(time: Number): string {
```
Comment (calbach) [inline ui/src/app/utils/dates.tsx:7]: `Number` is a wrapper object, probably this type should instead be `number`

File: ui/src/app/utils/dates.tsx [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  const date = new Date(Number(time));
```
Comment (calbach) [inline ui/src/app/utils/dates.tsx:8]: Existing issue: `Number(time)` is interesting; seems like it should be unnecessary given the type. If it is needed for some reason, a comment would be good.

File: ui/src/app/utils/dates.tsx [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+export function displayDateWithoutHours(time: number): string {
+  const date = new Date(time);
+  // datetime formatting to slice off weekday and exact time
+  return date.toLocaleString('en-us', {month: 'short', day: 'numeric', year: 'numeric'});
+}
+
+export function isDateValid(date: Date): boolean {
+  return date !== null && typeof date === 'object' && !isBlank(date.toString()) && !isNaN(date.valueOf());
```
Comment (calbach) [inline ui/src/app/utils/dates.tsx:24]: nit: it seems inconsistent to me to check for null but not undefined. That said - I think `typeof date` should already cover both cases

**Review by calbach — APPROVED** [review]

File: ui/src/app/pages/homepage/data-access-requirements.tsx [hunk 8]
```diff
@@ -732,7 +733,7 @@ const ControlledTierStep = (props: {enabled: boolean, text: String}) => {
 
 const DuccCard = (props: {
       profile: Profile, activeModule: AccessModule, clickableModules: AccessModule[],
-      spinnerProps: WithSpinnerOverlayProps, stepNumber: Number
+      spinnerProps: WithSpinnerOverlayProps, stepNumber: number
```
Comment (jmthibault79) [inline ui/src/app/pages/homepage/data-access-requirements.tsx:736]: fixing this existing Number issue as well.  Tested new version.

File: ui/src/app/utils/dates.tsx (hunk 6 above)
Comment (jmthibault79) [reply ui/src/app/utils/dates.tsx]: removed and tested

**Pull request merged**